plugins/inputs/t128_graphql/t128_graphql.go

```diff
@@ -0,0 +1,284 @@
+		return
+	}
+
+	//decode json
+	jsonParsed, err := gabs.ParseJSON(buf.Bytes())
+	if err != nil {
+		acc.AddError(fmt.Errorf("invalid json response for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	jsonEntryPoint := plugin.BuildJsonEntryPoint()
+	jsonObj, err := jsonParsed.JSONPointer(jsonEntryPoint)
+	if err != nil {
+		template := "unexpected response for metric %s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+
+	//update acc
+	jsonChildren, err := jsonObj.Children()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to iterate on response nodes for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	//TODO: allow nested fields/tags
+	for _, child := range jsonChildren {
+		node := child.Data().(map[string]interface{})
+		fields := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		//allow integer conversion?
+		//TODO: fully implement multiple fields
+		for fieldRenamed, fieldName := range plugin.Fields {
+			if isNil(node[fieldName]) {
+				acc.AddError(fmt.Errorf("found empty data for metric %s: field %s", plugin.CollectorName, fieldName))
+				continue
+			}
+			fields[fieldRenamed] = node[fieldName]
+
+			for tagRenamed, tagName := range plugin.Tags {
+				if isNil(node[tagName]) {
+					tags[tagRenamed] = tagRenamed
+					continue
+				}
+				//gabs module converts data to either string or float64
+				strTag, ok := node[tagName].(string)
+				if !ok {
+					floatTag := node[tagName].(float64)
+					strTag = strconv.FormatFloat(floatTag, 'f', -1, 64)
+				}
+				tags[tagRenamed] = strTag
+			}
+
+			acc.AddFields(
+				plugin.CollectorName,
+				fields,
+				tags,
+				timestamp,
+			)
+		}
+	}
+}
+
+func (plugin *T128GraphQL) createRequest() (*http.Request, error) {
+	content := struct {
+		Query string `json:"query,omitempty"`
+	}{
+		plugin.Query,
+	}
+
+	//inject env config?
+	body, err := json.Marshal(content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request body for query '%s': %w", plugin.Query, err)
+	}
+
+	url := plugin.BaseURL + "api/v1/graphql"
+	request, err := http.NewRequest("POST", url, bytes.NewReader(body))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request for query '%s': %w", plugin.Query, err)
+	}
+
+	request.Header.Add("Content-Type", "application/json")
+
+	return request, nil
+}
+
+func (plugin *T128GraphQL) BuildJsonEntryPoint() string {
+	path := "/data/"
+	pathElements := strings.Split(plugin.EntryPoint, "/")
+	for idx, element := range pathElements {
+		bracketIdx := strings.Index(element, "[")
+		if bracketIdx > 0 {
+			path += element[:bracketIdx] + "/"
+		} else {
+			if idx < len(pathElements)-2 {
+				path += element + "/0/"
+			} else {
+				path += element + "/"
+			}
+		}
+	}
+	path = strings.TrimRight(path, "/")
+	return path
+}
+
+func (plugin *T128GraphQL) DecodeAndReportJsonErrors(response []byte, acc telegraf.Accumulator, template string) {
```

Comment:
Let's make this function return a list of errors and then we can add the callsite error message to each at the callsite and then add it to the accumulator. Also then this could become a free function since we could inject the collectorname at the callsite

Reply:
- [ ] return errors and add to acc at callsite